src/tools/tools/process.fs

```diff
@@ -1171,10 +1171,130 @@ type internal ThreadPoolWait internal (id:int) as this =
         member x.Dispose() = 
             CleanUp.Current.CleanUpAll()
             GC.SuppressFinalize(x)
-            
-
-            
-
+          
+/// ThreadPoolWait.RegisterWaitForSingleObject, with capability:
+/// 1. To remove a particular wait handle. 
+/// 2. To show all wait handles queued in the system, 
+/// and 3. To show long continuation (not normal). 
+type internal ThreadPoolWait() = 
+    /// Maximum wait for the RegisteredWaitHandle to show up
+    static member val MaxSpinMs = 200 with get, set
+    static member val ActiveThreadPools = ConcurrentDictionary<_,_>() with get
+    static member val WaitHandleCollection = ConcurrentDictionary<_,ConcurrentDictionary<_,_>>() with get
+    static member RegisterThreadPool (name:string) = 
+        Logger.LogF( LogLevel.MediumVerbose, ( fun _ -> sprintf "ThreadPoolWait, register wait pool %s ............" name ))
+        ThreadPoolWait.ActiveThreadPools.Item( name ) <- DateTime.UtcNow
+    static member UnregisterThreadPool (name:string) = 
+        Logger.LogF( LogLevel.MildVerbose, ( fun _ -> sprintf "ThreadPoolWait, unregister wait pool %s ............" name ))
+        ThreadPoolWait.ActiveThreadPools.TryRemove( name ) |> ignore
+        if ThreadPoolWait.ActiveThreadPools.IsEmpty then 
+            if ( !(ThreadPoolWaitDeprecated.NumberOfWaitingThreads) > 0 ) then 
+                Logger.LogF( LogLevel.MildVerbose, ( fun _ -> sprintf "ThreadPoolWait, threadpool %s is the last registered threadpools, try terminating all" name ))
+            // Termniate all Wait thread. 
+            ThreadPoolWait.TryTerminateAll()
+    static member WaitForHandle (infoFunc: unit-> string) (handle:WaitHandle) (continuation:unit->unit) (unblockHandle:ManualResetEvent) =
+        if handle.WaitOne(0) then 
+            continuation() 
+            if Utils.IsNotNull unblockHandle then 
+                // If there is an unblock handle, set it. 
+                unblockHandle.Set() |> ignore  
+        else
+            let continuationObject = continuation :> Object
+            let rwh = ThreadPool.RegisterWaitForSingleObject( handle, new WaitOrTimerCallback(ThreadPoolWait.CallBack), (handle,continuationObject) , -1, true )
+            let handleCollection = ThreadPoolWait.WaitHandleCollection.GetOrAdd( handle, fun _ -> ConcurrentDictionary())
+            handleCollection.Item(continuationObject) <- ( rwh, infoFunc, ref (continuation, unblockHandle), ref DateTime.MinValue.Ticks )
+    static member private CallBack( state: Object ) (timeout:bool) = 
```

Comment:
the callback should take cancellation token as a input parameter, so it can do cooperative cancellation.